python/lsst/daf/butler/registry/_remoteRegistry.py

```diff
@@ -0,0 +1,431 @@
+
+    def __init__(self, server_uri: ButlerURI, defaults: RegistryDefaults, writeable: bool):
+        self._db = server_uri
+        self._defaults = defaults
+
+        # All PUT calls should be short-circuited if not writeable.
+        self._writeable = writeable
+
+        self._dimensions: Optional[DimensionUniverse] = None
+
+        # Does each API need to be sent the defaults so that the server
+        # can use specific defaults each time?
+
+        # Storage class information should be pulled from server.
+        # Dimensions should be pulled from server.
+
+    def __str__(self) -> str:
+        return str(self._db)
+
+    def __repr__(self) -> str:
+        return f"RemoteRegistry({self._db!r}, {self.dimensions!r})"
+
+    def isWriteable(self) -> bool:
+        # Docstring inherited from lsst.daf.butler.registry.Registry
+        # Can be used to prevent any PUTs to server
+        return self._writeable
+
+    def copy(self, defaults: Optional[RegistryDefaults] = None) -> Registry:
+        # Docstring inherited from lsst.daf.butler.registry.Registry
+        if defaults is None:
+            # No need to copy, because `RegistryDefaults` is immutable; we
+            # effectively copy on write.
+            defaults = self.defaults
+        return type(self)(self._db, defaults, self.isWriteable())
+
+    @property
+    def dimensions(self) -> DimensionUniverse:
+        # Docstring inherited from lsst.daf.butler.registry.Registry
+        if self._dimensions is not None:
+            return self._dimensions
+
+        # Access /dimensions.json on server and cache it locally.
+        try:
+            response = httpx.get(str(self._db.join("universe")))
+            response.raise_for_status()
+        except httpx.RequestError as e:
+            print(f"An error occurred while requesting {e.request.url!r}.")
+            raise e
+        except httpx.HTTPStatusError as e:
+            print(f"Error response {e.response.status_code} while requesting {e.request.url!r}.")
+            raise e
+
+        config = DimensionConfig.fromString(response.text, format="json")
+        self._dimensions = DimensionUniverse(config)
+        return self._dimensions
+
+    def refresh(self) -> None:
+        # Docstring inherited from lsst.daf.butler.registry.Registry
+
+        # Need to determine what to refresh.
+        # Might need a server method to return all the DatasetTypes up front.
+        pass
+
+    # def transaction():
+    #    What does a transaction mean for the client? The server is running
+    #    in a transaction but if the client wants one transaction that
+    #    involves two server methods how does that work? The server will have
+    #    already closed the transaction.
+    @contextlib.contextmanager
+    def transaction(self, *, savepoint: bool = False) -> Iterator[None]:
+        # No-op for now
+        try:
+            yield
+        except BaseException:
+            raise
```

Comment:
I had it be a no-op so that it would let the client experiment with doing writes on the server without having to solve the hard problem.  Another way to look at it is that this might lead to us getting in to trouble later on when we do want to think about transaction handling.  I'll change it to raise for the moment.